### ISSUE TITLE
Tutorials' erratum

### ISSUE THREAD

**Issue by martricks** [jaxleyverse/jaxley#720]
Dear Jaxley community,

I am an electrophysiologist and a programmer now at the postdoc stage. I wrote all the modeling part of my pre-doctoral work in the NEURON platform (hoc). Guys, what you’ve built and what you are currently building is simply great, I truly hope that it becomes the gold standard for compartmentalized modeling. Parallelizing is a decisive advantage, and the other NEURON-like functionalities can be implemented more elegantly with fewer lines of code, not to say that writing python instead of hoc just kicks butt.
Now to the point, I have followed your basic and advanced tutorials, and I have found a few things here and there that you might consider changing:

**Typos**
1. Tutorials > Speeding up simulations > Stimulus sweeps > 1st line:
“sweeep” with 3 “e” instead of 2
2. Tutorials > Training biophysical models > Defining parameter transformations > 2nd sentence:
“acocrdingly” …accordingly
3. Advanced tutorials > Key concepts in Jaxley > Channels > 2nd paragraph:
“Views come in handy, as it allows” …as they allow
4. Advanced tutorials > Customizing synaptic parameters > Example2 – Setting parameters given pre- and post- synaptic cell indices > 2nd paragraph:
“call ‘.query()’ to twice to filter” …call ‘.query()’, twice (without “to”)

**Broken links**
1. Tutorials > Building ion channels and synapses > final words in the last paragraph of the section “Your own channel”:
“this tutorial” redirects to an error-404 page
2. Tutorials > Building ion channels and synapses > the 1st link in the very last paragraph of the page:
“this tutorial” redirects to an error-404 page

**Code integrity**

- Tutorials > Training biophysical models > Using checkpointing: line: `jx.integrate(..., checkpoint_inds=checkpoints)`
For the script to properly run:
1st  `checkpoint_inds` needs to be changed to `checkpoint_lengths`
2nd  `t_max=10.0` needs to be changed to `t_max=5.0`
- Advanced tutorials > Simplified models > How to implement your own simplified models > declaration of `class Fire(Channel)` > method `update_states()`: in the 4th line (the one just before ‘return’)
For the script to properly run, `jax.lax.select()` needs to be called with `jnp.select()` because `jax.numpy` has initially been imported as `jnp`. 

**Style**
Rather personal opinions -take them with skepticism- most of them about the section: Tutorials > Training biophysical models
- Parameter’s score is assigned in an unconventional way which might lead to confusion. The array ‘labels’ first is logical, then is converted to float, afterwards the array ‘losses’ is computed by subtracting ‘labels’ from ‘predictions’ which is “roughly” normalized to 1.
- The blocks of code “Writing a dataloader” & “Training loop” are sparsely commented and this might give too much freedom of interpretation to the inexperienced reader.
- The blocks of code “Using checkpointing” & “Defining parameter transformations” are prerequisites to the blocks “Simulating the network” & “Defining a loss function” however the latter appear first in the tutorial: I would personally stick to the operational sequence when writing the tutorial’s story line, for two main reasons, (1) it might help the inexperienced reader with internalizing the logic chain to be followed when writing code, (2) it’s nice when the code of a tutorial in its original sequence can be copy-pasted into a notebook and it runs flawlessly straight ahead (which isn’t the case currently)
- It might be worth considering the option of a much simpler -but perhaps more realistic- fitting paradigm: the Log Likelihood Function 

```
# DEFINE A RECTIFYING NEURON
# -Barely responding (exponential dynamics) when stimulus "x" < threshold "w0"
# -Responding proportionally (linear transform "w") when stimulus "x" > "w0"
def invLinkFun(X, w):
    z = np.dot(X, w)            #fill "z" with the results from multiplying every value of stimulus "x" by the transform "w"
    mu = np.zeros(z.shape)      #preallocate array "mu" with the same size & structure as "z" but filled with zeros
    mu[z<0] = np.exp(z[z<0])    #if an element in "z" is <0 let the corresponding "mu" element be exponentially smaller
    mu[z>=0] = z[z>=0] + 1      #if an element in "z" is >=0 leave the corresponding "mu" element as linearly transformed & +1
    return mu

# SET INITIAL PARAMETERS
n = 200;            #set the number of data points (resolution/finesse)
wTrue = [-1, 0.5]   #define the vector containing the main parameters threshold "w0=-1" and linear transform "w=0.5"

# GENERATE AN ARTIFICIAL STIMULUS & ADD POISSON SPIKING
x = np.random.rand(n)*30 - 15   #generate an array of "n" elements randomly varying within a range of 30 (∓15)
x = np.sort(x)                  #sort in ascending order the previously generated values
X = np.stack([-np.ones(n),x]).T #pile the previous array on top of an equally sized array filled with ones & align vertically
mu = invLinkFun(X, wTrue)       #feed the neuron with the stimulus "X" + set its threshold "w0" & linear "w" transform values
y = np.random.poisson(mu)       #add poisson nature to the firing of the neuron

# FIND THE BEST COMBINATION OF PARAMETER VALUES "w0" & "w" TO FIT EXPERIMENTAL DATA
# -The probability of experimentally observed spike rate given stimulus "x" and params "w"&"w0" is p(y|x,w0,w)=[f(z)^y exp(-f(z))]/y!
# -Likelihood is the same probability but for all spike trains observed: a large summation for multiple instances of the previous probability
# -Logarithms turn products into sums, the log-likelihood simplifying the expression into logLik(w0,w1)=sum[y*log(f(z))-f(z)]

logLikFun = lambda X, w, y: np.sum(y*np.log(invLinkFun(X, w))-invLinkFun(X, w)) #declare log-likelihood as a lambda function

# DEFINE PARAMETERS' SPACE
#matrix containing 2 arrays: the 1st contains 3 modified versions of parameter "w0", the 2nd array contains the same thing for parameter "w"
wEx = np.array([[wTrue[0]-0.9, wTrue[0], wTrue[0]+0.9], [wTrue[1]+0.3, wTrue[1], wTrue[1]-0.3]])
nGrid = 51  #finesse of the 2D-grid representing the parameters' space w0-w with all possible values in range w0{-2 to 0} / w{0.2 to 0.8}
W0, W1 = np.meshgrid(np.linspace(wTrue[0]-1, wTrue[0]+1, nGrid), np.linspace(wTrue[1]-0.5, wTrue[1]+0.5, nGrid))  #get the 2D-grid

# EVALUATE THE log-LIKELIHOOD FUNCTION FOR ALL COMBINATIONS OF PARAMETERS w0-w OVER THE RANGE DEFINED ABOVE: w0{-2 to 0} / w{0.2 to 0.8}
llVals = np.zeros([nGrid, nGrid])             #2D-array of zeroes with size=51x51 so as to fit into the previously obtained 2D-grid
for i in range(nGrid):                        #nested-loop to iterate sequentially across the 2 dimensions of the array
    for j in range(nGrid):
        wTmp = np.array([W0[i, j], W1[i, j]]) #temporary array containing a unique combination w0-w to be overwritten each iteration
        llVals[i, j] = logLikFun(X, wTmp, y)  #evaluate the log-likelihood function with one unique w0-w combination each iteration
                                              #sequentially fill each empty element of the above preallocated 2D-array with each log-likelihood result
```

That’s it. I hope you find some of my comments useful and not exceedingly annoying. It will always be a pleasure for me to contribute to Jaxley as far as I manage to find some available time.
Thanks a lot for your patience and your attention.
All the best,
Marti
